### PR TITLE
Initial KafkaSinkCluster

### PR DESCRIPTION
**Pull request by rukai**
Completes the first chunk of work listed in https://github.com/shotover/shotover-proxy/issues/1214

Because the transform lacks routing it can only be used with a 1 node kafka cluster.
This still brings some value as we can host shotover on a different IP which KafkaSinkSingle does not allow.

The logic in the new `send_requests` method has been pulled out of the two sinks.

Introduces the xxhash-rust crate under the BSL license which should pose no issue as its just a more lenient version of the MIT license.

### CODE REVIEW

**Comment by benbromhead** [comment]
BSL is good. 

Personally I like to just use the ASF categories as described in https://www.apache.org/legal/resolved.html to mentally check compatibility without too much brain power (we can obviously make decisions different to the ASF, but it's a good starting position).

File: shotover/src/transforms/kafka/sink_cluster.rs [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+            )
+            .await?;
+            let (rx, tx) = tcp_stream.into_split();
+            self.outbound = Some(spawn_read_write_tasks(&codec, rx, tx));
+        }
+
+        let outbound = self.outbound.as_mut().unwrap();
+        let responses = send_requests(message_wrapper.messages, outbound)?;
+
+        // TODO: since kafka will never send requests out of order I wonder if it would be faster to use an mpsc instead of a oneshot or maybe just directly run the sending/receiving here?
```
Comment (benbromhead) [inline shotover/src/transforms/kafka/sink_cluster.rs:137]: Sounds like a good idea worth some investigation.

**Review by benbromhead — APPROVED** [review]

File: shotover-proxy/tests/test-configs/kafka/cluster/docker-compose.yaml [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+      - KAFKA_CFG_NODE_ID=0
+    volumes:
+      - type: tmpfs
+        target: /bitnami/kafka
+  kafka1:
+    networks:
+      cluster_subnet:
+        ipv4_address: 172.16.1.3
+    image: 'bitnami/kafka:3.4.0-debian-11-r22'
+    environment:
```
Comment (conorbros) [inline shotover-proxy/tests/test-configs/kafka/cluster/docker-compose.yaml:32]: can use docker anchor for these ones

File: shotover/src/transforms/kafka/sink_cluster.rs [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+
+fn hash_address(host: &str, port: i32) -> u64 {
+    let mut hasher = xxhash_rust::xxh3::Xxh3::new();
+    hasher.write(host.as_bytes());
+    hasher.write(&port.to_be_bytes());
+    hasher.finish()
+}
+
+fn rewrite_address(shotover_nodes: &[KafkaAddress], host: &mut StrBytes, port: &mut i32) {
+    //tracing::info!("rewriting {} {}", host, port);
```
Comment (conorbros) [inline shotover/src/transforms/kafka/sink_cluster.rs:224]: :p

File: shotover/src/transforms/kafka/sink_cluster.rs [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+        // TODO: since kafka will never send requests out of order I wonder if it would be faster to use an mpsc instead of a oneshot or maybe just directly run the sending/receiving here?
+        let mut responses = if let Some(read_timeout) = self.read_timeout {
+            timeout(read_timeout, read_responses(responses)).await?
+        } else {
+            read_responses(responses).await
+        }?;
+
+        // Rewrite responses to use shotovers port instead of kafkas port
+        for response in &mut responses {
+            tracing::info!("{:?}", response.to_high_level_string());
```
Comment (conorbros) [inline shotover/src/transforms/kafka/sink_cluster.rs:146]: :p

Comment (rukai) [reply shotover/src/transforms/kafka/sink_cluster.rs]: ah yes, thanks

File: shotover/src/transforms/kafka/common.rs [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+use tokio::sync::oneshot;
+
+pub fn send_requests(
+    messages: Vec<Message>,
+    outbound: &mut Connection,
+) -> Result<Vec<oneshot::Receiver<Response>>> {
+    messages
+        .into_iter()
+        .map(|mut message| {
+            let acks0 = if let Some(Frame::Kafka(KafkaFrame::Request {
```
Comment (conorbros) [inline shotover/src/transforms/kafka/common.rs:19]: Could you add a comment with a small explanation for this bit

**Review by conorbros — APPROVED** [review]
LGTM

**Pull request merged**